src/Avalonia.Base/Input/Platform/DataObjectToDataTransferItemWrapper.cs

```diff
@@ -19,13 +27,85 @@ internal sealed class DataObjectToDataTransferItemWrapper(
     protected override DataFormat[] ProvideFormats()
         => _formats;
 
-    protected override object? TryGetCore(DataFormat format)
+    protected override object? TryGetRawCore(DataFormat format)
     {
         var index = Array.IndexOf(Formats, format);
         if (index < 0)
             return null;
 
+        // We should never have DataFormat.File here, it's been handled by DataObjectToDataTransferWrapper.
+        Debug.Assert(!DataFormat.File.Equals(format));
+
         var formatString = _formatStrings[index];
-        return _dataObject.Get(formatString);
+        var data = _dataObject.Get(formatString);
+
+        if (DataFormat.Text.Equals(format))
+            return Convert.ToString(data) ?? string.Empty;
+
+        if (format is DataFormat<byte[]>)
+            return ConvertLegacyDataToBytes(format, data);
+
+        return null;
+    }
+
+    private static byte[]? ConvertLegacyDataToBytes(DataFormat format, object? data)
+    {
+        switch (data)
+        {
+            case null:
+                return null;
+
+            case byte[] bytes:
+                return bytes;
+
+            case Memory<byte> bytes:
+                return bytes.ToArray();
+
+            case string str:
+                return OperatingSystemEx.IsWindows() || OperatingSystemEx.IsMacOS() || OperatingSystemEx.IsIOS() ?
+                    Encoding.Unicode.GetBytes(str) :
+                    Encoding.UTF8.GetBytes(str);
+
+            case Stream stream:
+                var length = (int)(stream.Length - stream.Position);
+                var buffer = new byte[length];
+
+                stream.ReadExactly(buffer, 0, length);
+                return buffer;
+
+            default:
+                if (OperatingSystemEx.IsWindows())
+                {
+                    Logger.TryGet(LogEventLevel.Warning, LogArea.Win32Platform)?.Log(
+                        null,
+                        "Using BinaryFormatter to serialize data format {Format}. This won't be supported in Avalonia v12. Prefer passing a byte[] or Stream instead.",
+                        format);
+
+                    return SerializeUsingBinaryFormatter(data);
+                }
+
+                return null;
+        }
+    }
+
+    // TODO12: remove
```

Comment:
Just watch the TODO's - They can build up over time if not dealt with in a PR :)